File: Code/Websites/DanpheEMR/wwwroot/DanpheApp/src/app/inventory/shared/goods-receipt.model.ts
import {
  FormGroup,
  FormControl,
  Validators,
  FormBuilder,
} from '@angular/forms'
import { GoodsReceiptItems } from "./goods-receipt-item.model"
import { ENUM_GRItemCategory } from "../../shared/shared-enums";
import * as moment from 'moment/moment';
export class GoodsReceipt {
  public GoodsReceiptValidator: FormGroup = null;
  public VendorName: string = "";
  public VendorNo: string = "";
  public ContactAddress: string = "";
  public GRCategory: string = ENUM_GRItemCategory.Consumables;
  public IsCancel: boolean = false;
  AgingDays: number;
  SupplierId: number;
  length: number;
  public FromDate: string = "";
  public ToDate: string = "";
  public CancelRemarks: string = "";
  //added for Verification
  public IsVerificationEnabled: boolean = false;
  public VerifierList: any[] = [];
  public VerifierIds: string;
  public CurrentVerificationLevel: number;
  public CurrentVerificationLevelCount: any;
  public MaxVerificationLevel: any;
  public IsVerificationAllowed: boolean;
  public VerificationStatus: string;
  public GRStatus: string;
  public GoodsArrivalFiscalYearFormatted: string;
  public GoodsArrivalNo: number;
  public GoodsArrivalDate : string;
  public IMIRNo: number;
  public IMIRDate : string;
  constructor() {

    var _formBuilder = new FormBuilder();
    this.GoodsReceiptValidator = _formBuilder.group({
      //sanjit: 2Apr'20: GoodsReceiptDate somehow throws validation error when use with danphe-date-picker, so it is commented. 
      // 'GoodsReceiptDate': ['', Validators.compose([Validators.required])],
      'BillNo': ['', Validators.compose([Validators.required])],
      // 'DonationId': ['', Validators.compose([Validators.required])],
      'VendorId': ['', Validators.compose([Validators.required])],
      'PaymentMode': ['', Validators.compose([Validators.required])],
    });
  }

  public IsDirty(fieldName): boolean {
    if (fieldName == undefined)
      return this.GoodsReceiptValidator.dirty;
    else
      return this.GoodsReceiptValidator.controls[fieldName].dirty;
  }


  public IsValid(): boolean { if (this.GoodsReceiptValidator.valid) { return true; } else { return false; } } public IsValidCheck(fieldName, validator): boolean {
    if (fieldName == undefined) {
      return this.GoodsReceiptValidator.valid;
    }
    else
      return !(this.GoodsReceiptValidator.hasError(validator, fieldName));
  }

  dateValidator(control: FormControl): { [key: string]: boolean } {

    //get current date, month and time
    var currDate = moment().format('YYYY-MM-DD');
    //if positive then selected date is of future else it of the past || selected year can't be of future
    if (control.value) {
      if ((moment(control.value).diff(currDate) > 0)
        || (moment(control.value).diff(currDate, 'years') > 1))
        return { 'wrongDate': true };
    }


    else
      return { 'wrongDate': true };

  }
  public GoodsReceiptID: number = 0;
  public GoodsReceiptNo: number = 0;
  public GoodsReceiptDate: string = moment().format('YYYY-MM-DD');
  public VendorBillDate: string = moment().format('YYYY-MM-DD');
  public PurchaseOrderId: number = null;
  public TotalAmount: number = 0;
  public Remarks: string = "";
  public CreatedBy: number = 0;
  public CreatedOn: string = null;
  public VendorId: number = 0;
  public FiscalYearId: number = 0;
  public CurrentFiscalYear: string = "";
  public BillNo: string = null;
  public ReceivedDate: string = moment().format('YYYY-MM-DD HH:MM:ss');
  public ReceiptNo: string = null;
  public OrderDate: string = null;

  //SubTotal and VATTotal only for caculation and show purpose
  public SubTotal: number = 0;
  public VATTotal: number = 0;
  public CcCharge: number = 0;
  public Discount: number = 0;
  public DiscountAmount: number = 0;
  public TDSRate: number = 0;
  public TDSAmount: number = 0;
  public TotalWithTDS: number = 0;
  public PrintCount: number = 0;
  //public SelectedItem: any = null;//sud:11Apr'20-- this was misused.. hence removing.
  public CreditPeriod: number = 0;
  public PaymentMode: string;
  public OtherCharges: number = 0;
  public GoodsReceiptItem: Array<GoodsReceiptItems> = new Array<GoodsReceiptItems>();

  //for other charges
  public InsuranceCharge: number = 0;
  public CarriageFreightCharge: number = 0;
  public PackingCharge: number = 0;
  public TransportCourierCharge: number = 0;
  public OtherCharge: number = 0;

  //for edit option
  public IsTransferredToACC: boolean = false;
  public ModifiedBy: number = null;
  public ModifiedOn: Date = null;

  //for Donations purpose tilganga hospital in INV_TXN_FixedAssetStock table 
  public DonationId: any;
}

File: Code/Components/DanpheEMR.ServerModel/InventoryModels/InventoryReportModel/GoodsReceiptModel.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace DanpheEMR.ServerModel.InventoryModels.InventoryReportModel
{
    public class GoodsReceiptModel
    {
        public bool? IsCancel { get; set; }

        [Key]
        public int GoodsReceiptID { get; set; }

        public DateTime? GoodsReceiptDate { get; set; }
        public int? PurchaseOrderId { get; set; }
        public decimal TotalAmount { get; set; }
        public string Remarks { get; set; }
        public int? CreatedBy { get; set; }
        public DateTime? CreatedOn { get; set; }
        public int VendorId { get; set; }
        public string BillNo { get; set; }
        public DateTime? ReceivedDate { get; set; }
        public string ReceiptNo { get; set; }
        public DateTime? OrderDate { get; set; }

        public bool? IsTransferredToACC { get; set; }
        //SubTotal and VATTotal only for caculation and show purpose
        public decimal SubTotal { get; set; }
        public decimal VATTotal { get; set; }
        public decimal? TDSRate { get; set; }
        public decimal? TDSAmount { get; set; }
        public decimal? TotalWithTDS { get; set; }
        public decimal CcCharge { get; set; }
        public double Discount { get; set; }
        public decimal DiscountAmount { get; set; }
        public int PrintCount { get; set; }
        public int CreditPeriod { get; set; }
        public string PaymentMode { get; set; }
        public decimal? OtherCharges { get; set; }
        //for other charges
        public decimal? InsuranceCharge { get; set; }
        public decimal? CarriageFreightCharge { get; set; }
        public decimal? PackingCharge { get; set; }
        public decimal? TransportCourierCharge { get; set; }
        public decimal? OtherCharge { get; set; }
        public virtual List<GoodsReceiptItemsModel> GoodsReceiptItem { get; set; }
        public int? ModifiedBy { get; set; }
        public DateTime? ModifiedOn { get; set; }

        //public VendorMasterModel Vendor { get; set; }
        //  public  decimal? TotalWithoutTDS { get; set; }

        [NotMapped]
        public int VoucherId { get; set; }
        [NotMapped]
        public string VoucherName { get; set; }
        [NotMapped]
        public string TransactionType { get; set; }
        [NotMapped]
        public decimal SalesAmount { get; set; }
        [NotMapped]
        public decimal VATAmount { get; set; }
        [NotMapped]
        public List<int> ReferenceIds { get; set; }
        [NotMapped]
        public string Type { get; set; }
        [NotMapped]
        public string VendorName { get; set; }
        [NotMapped]
        public int ItemId { get; set; }
        [NotMapped]
        public string ItemName { get; set; }
        [NotMapped]
        public string ContactNo { get; set; }
        [NotMapped]
        public List<SyncBillingAccountingModel> BillSyncs { get; set; }

        [NotMapped]
        public DateTime? PoDate { get; set; }
        public string CancelRemarks { get; set; }
        public DateTime? CancelledOn { get; set; }
        public int? CancelledBy { get; set; }
        public int? GoodsReceiptNo { get; set; }
    }
}
